app/lib/discourse/single_sign_on.rb

```diff
@@ -0,0 +1,108 @@
+module Discourse
+  class SingleSignOn
+    #
+    # Conveniently cargo culted from https://github.com/discourse/discourse/blob/master/lib/single_sign_on.rb
+    # with some very small changes to #parse since we're passing data in via json and not query params
+
+    ACCESSORS = [:nonce, :name, :username, :email, :avatar_url, :avatar_force_update, :require_activation,
+                 :bio, :external_id, :return_sso_url, :admin, :moderator, :suppress_welcome_message, :title,
+                 :add_groups, :remove_groups]
+    FIXNUMS = []
+    BOOLS = [:avatar_force_update, :admin, :moderator, :require_activation, :suppress_welcome_message]
+    NONCE_EXPIRY_TIME = 10.minutes
+
+    SECRET = OperationCode.fetch_secret_with name: :discourse_sso_secret
+
+    attr_accessor(*ACCESSORS)
+    attr_accessor :sso_secret, :sso_url
+
+    def self.sso_secret
+      raise RuntimeError, "sso_secret not implemented on class, be sure to set it on instance"
+    end
+
+    def self.sso_url
+      raise RuntimeError, "sso_url not implemented on class, be sure to set it on instance"
+    end
+
+    def self.parse(payload, sso_secret = nil)
```

Comment:
I don't love this code either but it's the reference implementation from discourse: https://github.com/discourse/discourse/blob/master/lib/single_sign_on.rb

I opted out of making any changes (except for the necessary ones for our environment) to make any troubleshooting we might need to do easier. I dropped a comment here noting this: https://github.com/OperationCode/operationcode_backend/blob/3b98c8371a351a2b20f3727ebc9bfbcc4b7e6001/app/lib/discourse/single_sign_on.rb#L4